Provider v0.84.0 crashes when running snowflake_grant_database_role

### Terraform CLI and Provider Versions

`Terraform v1.6.4 on darwin_arm64`

Provider version v0.84.0

### Terraform Configuration

```terraform
resource "snowflake_database_role" "database_role" {
  database = "DB_TEST_DB_DEV"
  name     = "TEST_DATABASE_ROLE_02"
}

resource "snowflake_database_role" "parent_database_role" {
  database = "DB_TEST_DB_DEV"
  name     = "TEST_DATABASE_ROLE_01"
}

resource "snowflake_grant_database_role" "g" {
  database_role_name        = "\"DB_TEST_DB_DEV\".\"${snowflake_database_role.database_role.name}\""
  parent_database_role_name = "\"DB_TEST_DB_DEV\".\"${snowflake_database_role.parent_database_role.name}\""
}
```


### Expected Behavior

Database role TEST_DATABASE_ROLE_02 granted to TEST_DATABASE_ROLE_01

### Actual Behavior

```
panic: runtime error: index out of range [1] with length 1

goroutine 64 [running]:
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.ReadGrantDatabaseRole(0x14000878780, {0x1065e0960?, 0x14001016d00?})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/grant_database_role.go:184 +0x668
github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources.CreateGrantDatabaseRole(0x0?, {0x1065e0960?, 0x14001016d00})
	github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/resources/grant_database_role.go:131 +0x758
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x106622bc8?, {0x106622bc8?, 0x14000a30990?}, 0xd?, {0x1065e0960?, 0x14001016d00?})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:766 +0x134
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x14000b310a0, {0x106622bc8, 0x14000a30990}, 0x1400086eb60, 0x14000878600, {0x1065e0960, 0x14001016d00})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/resource.go:909 +0x86c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x140009fc648, {0x106622bc8?, 0x14000a307e0?}, 0x14000642f50)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.31.0/helper/schema/grpc_provider.go:1074 +0xb08
github.com/hashicorp/terraform-plugin-mux/tf5to6server.v5tov6Server.ApplyResourceChange({{0x106637bd8?, 0x140009fc648?}}, {0x106622bc8, 0x14000a307e0}, 0x0?)
	github.com/hashicorp/terraform-plugin-mux@v0.13.0/tf5to6server/tf5to6server.go:37 +0x58
github.com/hashicorp/terraform-plugin-mux/tf6muxserver.(*muxServer).ApplyResourceChange(0x106622c00?, {0x106622bc8?, 0x14000a304e0?}, 0x14000642f00)
	github.com/hashicorp/terraform-plugin-mux@v0.13.0/tf6muxserver/mux_server_ApplyResourceChange.go:36 +0x184
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0x14000bee000, {0x106622bc8?, 0x1400086bcb0?}, 0x140008724d0)
	github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov6/tf6server/server.go:857 +0x3c0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x106592a40?, 0x14000bee000}, {0x106622bc8, 0x1400086bcb0}, 0x14000878380, 0x0)
	github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:503 +0x164
google.golang.org/grpc.(*Server).processUnaryRPC(0x140000cad20, {0x106622bc8, 0x1400086bc20}, {0x1066331e0, 0x14000102ea0}, 0x1400086d8c0, 0x14000bdf2c0, 0x1074aba78, 0x0)
	google.golang.org/grpc@v1.60.0/server.go:1372 +0xb8c
google.golang.org/grpc.(*Server).handleStream(0x140000cad20, {0x1066331e0, 0x14000102ea0}, 0x1400086d8c0)
	google.golang.org/grpc@v1.60.0/server.go:1783 +0xc4c
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.60.0/server.go:1016 +0x5c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 52
	google.golang.org/grpc@v1.60.0/server.go:1027 +0x138

Error: The terraform-provider-snowflake_v0.84.0 plugin crashed!
```

### Steps to Reproduce

1. `terraform apply`


### How much impact is this issue causing?

Medium

### Logs

_No response_

### Additional Information

```
2024-01-23T12:59:52.864Z [DEBUG] provider.terraform-provider-snowflake_v0.84.0: 2024/01/23 12:59:52 [DEBUG] sql-conn-exec: [query GRANT DATABASE ROLE "DB_TEST_DB_DEV"."TEST_DATABASE_ROLE_02" TO DATABASE ROLE "DB_TEST_DB_DEV"."TEST_DATABASE_ROLE_01" err <nil> duration 224.328125ms args {}] ()
2024-01-23T12:59:53.096Z [DEBUG] provider.terraform-provider-snowflake_v0.84.0: 2024/01/23 12:59:53 [DEBUG] sql-conn-query: [query SHOW GRANTS OF DATABASE ROLE "DB_TEST_DB_DEV"."TEST_DATABASE_ROLE_02" err <nil> duration 231.756667ms args {}] ()
2024-01-23T12:59:53.099Z [DEBUG] provider.terraform-provider-snowflake_v0.84.0: panic: runtime error: index out of range [1] with length 1
```

--- a/pkg/resources/grant_database_role.go
+++ b/pkg/resources/grant_database_role.go
@@ -179,6 +179,10 @@ func ReadGrantDatabaseRole(d *schema.ResourceData, meta interface{}) error {
 						parts = strings.Split(s, "\".")
 						if len(parts) < 2 {
 							parts = strings.Split(s, ".\"")
+							if len(parts) < 2 {
+								s = strings.Trim(s, "\"")
+								parts = strings.Split(s, ".")
+							}
 						}
 					}
 					s = sdk.NewDatabaseObjectIdentifier(parts[0], parts[1]).FullyQualifiedName()

--- a/pkg/resources/grant_database_role_acceptance_test.go
+++ b/pkg/resources/grant_database_role_acceptance_test.go
@@ -57,6 +57,39 @@ func TestAcc_GrantDatabaseRole_databaseRole(t *testing.T) {
 	})
 }
 
+func TestAcc_GrantDatabaseRole_issue2402(t *testing.T) {
+	databaseName := "DB_TEST_DB_DEV"
+	databaseRoleName := "TEST_DATABASE_ROLE_02"
+	parentDatabaseRoleName := "TEST_DATABASE_ROLE_01"
+	resourceName := "snowflake_grant_database_role.g"
+	m := func() map[string]config.Variable {
+		return map[string]config.Variable{
+			"database":                  config.StringVariable(databaseName),
+			"database_role_name":        config.StringVariable(databaseRoleName),
+			"parent_database_role_name": config.StringVariable(parentDatabaseRoleName),
+		}
+	}
+	resource.Test(t, resource.TestCase{
+		ProtoV6ProviderFactories: acc.TestAccProtoV6ProviderFactories,
+		PreCheck:                 func() { acc.TestAccPreCheck(t) },
+		TerraformVersionChecks: []tfversion.TerraformVersionCheck{
+			tfversion.RequireAbove(tfversion.Version1_5_0),
+		},
+		CheckDestroy: testAccCheckGrantDatabaseRoleDestroy,
+		Steps: []resource.TestStep{
+			{
+				ConfigDirectory: acc.ConfigurationDirectory("TestAcc_GrantDatabaseRole/issue2402"),
+				ConfigVariables: m(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "database_role_name", fmt.Sprintf(`"%v"."%v"`, databaseName, databaseRoleName)),
+					resource.TestCheckResourceAttr(resourceName, "parent_database_role_name", fmt.Sprintf(`"%v"."%v"`, databaseName, parentDatabaseRoleName)),
+					resource.TestCheckResourceAttr(resourceName, "id", fmt.Sprintf(`"%v"."%v"|DATABASE ROLE|"%v"."%v"`, databaseName, databaseRoleName, databaseName, parentDatabaseRoleName)),
+				),
+			},
+		},
+	})
+}
+
 func TestAcc_GrantDatabaseRole_accountRole(t *testing.T) {
 	databaseRoleName := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))
 	parentRoleName := strings.ToUpper(acctest.RandStringFromCharSet(10, acctest.CharSetAlpha))

--- a/pkg/resources/testdata/TestAcc_GrantDatabaseRole/issue2402/test.tf
+++ b/pkg/resources/testdata/TestAcc_GrantDatabaseRole/issue2402/test.tf
@@ -1,0 +1,18 @@
+resource "snowflake_database" "database" {
+  name = var.database
+}
+
+resource "snowflake_database_role" "database_role" {
+  database = snowflake_database.database.name
+  name     = var.database_role_name
+}
+
+resource "snowflake_database_role" "parent_database_role" {
+  database = snowflake_database.database.name
+  name     = var.parent_database_role_name
+}
+
+resource "snowflake_grant_database_role" "g" {
+  database_role_name        = "\"${snowflake_database.database.name}\".\"${snowflake_database_role.database_role.name}\""
+  parent_database_role_name = "\"${snowflake_database.database.name}\".\"${snowflake_database_role.parent_database_role.name}\""
+}

--- a/pkg/resources/testdata/TestAcc_GrantDatabaseRole/issue2402/variables.tf
+++ b/pkg/resources/testdata/TestAcc_GrantDatabaseRole/issue2402/variables.tf
@@ -1,0 +1,11 @@
+variable "database_role_name" {
+  type = string
+}
+
+variable "parent_database_role_name" {
+  type = string
+}
+
+variable "database" {
+  type = string
+}

